gemlog/gemlog.py

```diff
@@ -505,56 +505,133 @@ def MillisToTime(L):
     -------
     dict of dataframes
-
         - data: the analog readings and associated timings
         - metadata: datalogger metadata
         - gps: GPS timing and location values
     """
+    try:
+        from gemlog.parsers import parse_gemfile
+    except ImportError:
+        raise ImportError(
+            "gemlog's C-extensions are not available. Reinstall gemlog with "
+            "C-extensions to use this function."
+        )
+
+    # use cythonized reader file instead of pd.read_csv and slow string ops
+    values, types, millis = parse_gemfile(str(filename).encode('utf-8'))
+    if values.shape[0] == 0:
+        raise EmptyRawFile(filename)
+
+    df = pd.DataFrame(values, columns=range(2, 13))
+    # note that linetype has type bytes here, not str like in the pandas func
+    df['linetype'] = types
+    df['millis-sawtooth'] = millis
+    return _process_gemlog_data(df, offset)
+
+
+def _read_with_pandas(filename, offset=0):
     # skiprows is important so that the header doesn't force dtype=='object'
     # the C engine for pd.read_csv is fast but crashes sometimes. Use the python engine as a backup.
     try:
         df = pd.read_csv(filename, names=range(13), low_memory=False, skiprows=6)
-    except:
+    except Exception:
         df = pd.read_csv(filename, names=range(13), engine='python', skiprows=6, error_bad_lines = False, warn_bad_lines = False)
 
     if df.shape[0] == 0:
         raise EmptyRawFile(filename)
-    
+
     df['linetype'] = [value[0] for value in df[0]]
     df = df.loc[df.loc[:,'linetype'].isin(['D', 'M', 'G']), :]
     #df = df.loc[df['linetype'].isin(['D', 'M', 'G']), :]
     ## most of the runtime is before here
     # unroll the ms rollover sawtooth
     df['millis-sawtooth'] = np.where(df['linetype'] == 'D',df[0].str[1:],df[1]).astype(int)
+    return _process_gemlog_data(df, offset)
+
+
+def ReadGem_v0_9_single(filename, offset=0):
+    """
+    Read a Gem logfile.
+
+    Parameters
+    ----------
+    filename : str or pathlib.Path
+        Filepath of a file containing data to read.
+
+    offset : int, default 0
+        A timing offset to include in the millisecond timestamp values.
+
+    Returns
+    -------
+    dict of dataframes
+
+        - data: the analog readings and associated timings
+        - metadata: datalogger metadata
+        - gps: GPS timing and location values
+    """
+    # Try each of the three file readers in order of decreasing speed but
+    # probably increasing likelihood of success.
+
+    readers = [
+        _read_with_cython, _read_with_pandas, _slow_ReadGem_v0_9_single
+    ]
+    for reader in readers:
+        try:
+            return reader(filename, offset)
+        except (EmptyRawFile, FileNotFoundError):
+            # If the file is definitely not going to work, exit early and
+            # re-raise the exception that caused the problem
+            raise
+        except Exception:
+            pass
+
+    raise CorruptRawFile(filename)
```

Comment:
Is this the correct error to raise if none of the readers can handle the file?

Reply:
Yep.